zproject/backends.py

```diff
@@ -1003,6 +1009,31 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
             verified_emails.append(details["email"])
         return verified_emails
 
+class SAMLAuthBackend(SocialAuthMixin, SAMLAuth):
+    auth_backend_name = "SAML"
+
+    def auth_complete(self, *args: Any, **kwargs: Any) -> Optional[HttpResponse]:
+        """
+        Additional ugly wrapping on top of auth_complete in SocialAuthMixin.
+        If the /complete/saml endpoint is accessed, but RelayState isn't specified
+        in POST, django.utils.datastructures.MultiValueDictKeyError will be raised
+        inside python-social-auth saml module, which doesn't convert it into something
+        more reasonable.
+        If it is accessed with RelayState specified, but missing SAMLResponse,
+        OneLogin_Saml2_Error ends up raised.
+        We need to catch and handle those here.
```

Comment:
I think we should add a bit more context to this docstring explaining for someone who doesn't know SAML what these conditions are.  E.g. we should explain what "RelayState isn't specified in POST" means (like is that something that'd be caused by a bug, and previously provided an inscrutible error message?  That's sorta what this sounds like, but it's worth spelling out things like that explicitly as they'll be very helpful for anyone trying to understand this code in 6 months)

Reply:
Added an additional paragraph there:
```
        Both parameters should be present if the user came to /complete/saml/ through
        the IdP as intended. The errors can happen if someone simply types the endpoint into
        their browsers, or generally tries messing with it in some ways.
```